### PR TITLE
feat: Add external project add support to cmake

### PR DESCRIPTION
**Pull request by ddiakiteaneo**
Use externalproject_add instead of add_subdirectory to handle dependency 
Each module will us the install version of its dependency

### CODE REVIEW

File: ArmoniK.SDK.Client.Test/CMakeLists.txt [hunk 1]
```diff
@@ -1,31 +1,63 @@
 cmake_minimum_required(VERSION 3.22)
 set(PROJECT_NAME ArmoniK.SDK.Client.Test)
 set(NAMESPACE ArmoniK::Sdk::Client::Test)
+set(version 0.1.0)
+
 project(${PROJECT_NAME})
+set(CMAKE_INSTALL_INCLUDEDIR ${CMAKE_INSTALL_PREFIX}/include)
```
Comment (dbrasseur-aneo) [inline ArmoniK.SDK.Client.Test/CMakeLists.txt:7]: I don't think it's needed

File: ArmoniK.SDK.Client.Test/CMakeLists.txt [hunk 2]
```diff
@@ -1,31 +1,63 @@
 cmake_minimum_required(VERSION 3.22)
 set(PROJECT_NAME ArmoniK.SDK.Client.Test)
 set(NAMESPACE ArmoniK::Sdk::Client::Test)
+set(version 0.1.0)
```
Comment (dbrasseur-aneo) [inline ArmoniK.SDK.Client.Test/CMakeLists.txt:4]: I'd prefer it to be a variable to be set by the parent CMakeList (and give it as -DVERSION for example)

File: ArmoniK.SDK.Client.Test/CMakeLists.txt [hunk 3]
```diff
@@ -1,31 +1,63 @@
...
+
+if(IS_ALPINE EQUAL -1)
+    message(STATUS "Not found Alpine distribution " ${IS_ALPINE})
+else()
+    message(STATUS "Found Alpine distribution " ${IS_ALPINE})
+endif()
+
+if(CMAKE_BUILD_TYPE MATCHES DEBUG AND IS_ALPINE EQUAL -1)
+    target_link_options(${PROJECT_NAME} PRIVATE -fsanitize=undefined,address)
+endif()
```
Comment (dbrasseur-aneo) [inline ArmoniK.SDK.Client.Test/CMakeLists.txt:46]: Can be put in a Utils.cmake file and then this Utils.cmake can be included here with 
```include(Utils.cmake)```

File: ArmoniK.SDK.Client.Test/CMakeLists.txt [hunk 4]
```diff
@@ -1,31 +1,63 @@
...
         "${SOURCES_FILES_DIR}"
+        "$<BUILD_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>"
         )
 
 FILE(COPY appsettings.json DESTINATION "${PROJECT_BINARY_DIR}")
 
+if(POLICY CMP0135)
+	cmake_policy(SET CMP0135 NEW)
+	set(CMAKE_POLICY_DEFAULT_CMP0135 NEW)
+endif()
```
Comment (dbrasseur-aneo) [inline ArmoniK.SDK.Client.Test/CMakeLists.txt:59]:
```suggestion
if(POLICY CMP0135)
	cmake_policy(SET CMP0135 OLD)
endif()
```

File: ArmoniK.SDK.Client/CMakeLists.txt [hunk 5]
```diff
@@ -2,11 +2,16 @@ cmake_minimum_required(VERSION 3.22)
...
 set(NAMESPACE ArmoniK::Sdk::Client)
 project(${PROJECT_NAME})
+set(CMAKE_INSTALL_INCLUDEDIR ${CMAKE_INSTALL_PREFIX}/include)
+set(CMAKE_INSTALL_LIBDIR ${CMAKE_INSTALL_PREFIX}/lib)
 set(ARMONIK_INSTALL_INCLUDE_DIR ${CMAKE_INSTALL_INCLUDEDIR}/armonik/sdk/client)
 set(PROJECT_BUILD_DIR ${BUILD_DIR}/${PROJECT_NAME})
 
 set(CMAKE_CXX_STANDARD 17)
 
+set(VERSION 0.1.0)
```
Comment (dbrasseur-aneo) [inline ArmoniK.SDK.Client/CMakeLists.txt:12]: same remark

File: ArmoniK.SDK.DynamicWorker/CMakeLists.txt [hunk 6]
```diff
@@ -48,4 +71,4 @@ install(TARGETS ${PROJECT_NAME}
         RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
         ARCHIVE DESTINATION ${CMAKE_INSTALL_BINDIR})
 
-install(FILES appsettings.json DESTINATION ${CMAKE_INSTALL_BINDIR})
+install(FILES appsettings.json DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
```
Comment (dbrasseur-aneo) [inline ArmoniK.SDK.DynamicWorker/CMakeLists.txt:74]: It should mean the same thing

File: CMakeLists.txt [hunk 7]
```diff
@@ -19,8 +19,12 @@ list(APPEND CMAKE_PREFIX_PATH ${ARMONIK_API_DIR})
 if (WIN32)
     list(APPEND CMAKE_PREFIX_PATH "${CMAKE_SOURCE_DIR}/tools/win64/lib/cmake/grpc")
     list(APPEND CMAKE_PREFIX_PATH "${CMAKE_SOURCE_DIR}/tools/win64/cmake")
+    set(INSTALL_SDK_DIR ${CMAKE_SOURCE_DIR}/install)
     set(Protobuf_USE_STATIC_LIBS ON)
+else()
+    set(INSTALL_SDK_DIR "/app/install")
```
Comment (dbrasseur-aneo) [inline CMakeLists.txt:25]: Should not be hard coded. ${CMAKE_SOURCE_DIR}/install would be fine here I think

File: CMakeLists.txt [hunk 8]
```diff
@@ -30,60 +34,115 @@ include(CMakePackageConfigHelpers)
...
-            target_compile_options(${project_name_param} PRIVATE -Wall -Wextra -Wpedantic -fsanitize=undefined,address ${extra_param})
-        else ()
-            target_compile_options(${project_name_param} PRIVATE -Wall -Wextra -Wpedantic ${extra_param})
-        endif()
+    if(BUILD_WORKER)
+        ExternalProject_Add(ArmoniK.SDK.Worker
+            SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/ArmoniK.SDK.Worker"
+            BINARY_DIR "${CMAKE_CURRENT_SOURCE_DIR}/build/ArmoniK.SDK.Worker"
+            CMAKE_ARGS -DCMAKE_INSTALL_PREFIX:PATH=${INSTALL_SDK_DIR} -DCMAKE_PREFIX_PATH=${ARMONIK_API_DIR}
+            DEPENDS ArmoniK.SDK.Common
```
Comment (dbrasseur-aneo) [inline CMakeLists.txt:81]:
```suggestion
            DEPENDS ArmoniK.SDK.Common
```
Actually doesn't depend on it

File: CMakeLists.txt [hunk 9]
```diff
@@ -30,60 +34,115 @@ include(CMakePackageConfigHelpers)
...
+            STEP_TARGETS build
+        )
     endif()
+endif()
 
-    if(CMAKE_BUILD_TYPE MATCHES DEBUG AND IS_ALPINE EQUAL -1)
-        target_link_options(${project_name_param} PRIVATE -fsanitize=undefined,address ${extra_param})
+if(BUILD_DYNAMICWORKER)
+    If(NOT BUILD_WORKER)
+        find_package(ArmoniK.SDK.Worker CONFIG REQUIRED)
```
Comment (dbrasseur-aneo) [inline CMakeLists.txt:101]:
```suggestion
        find_package(ArmoniK.SDK.Common CONFIG REQUIRED)
        find_package(ArmoniK.SDK.Worker CONFIG REQUIRED)
```

File: CMakeLists.txt
Comment (dbrasseur-aneo) [inline CMakeLists.txt:1]: Step targets are different, not sure it impacts anything, or is relevant at all

**Review by dbrasseur-aneo — CHANGES_REQUESTED** [review]

File: ArmoniK.SDK.DynamicWorker/CMakeLists.txt (hunk 6 above)
Comment (ddiakiteaneo) [reply ArmoniK.SDK.DynamicWorker/CMakeLists.txt:53]: Yes, it is the same thing, but if the folder is not yet created CMake generation fails for FILES copies. This is why I choose the second option.

File: CMakeLists.txt (hunk 7 above)
Comment (ddiakiteaneo) [reply CMakeLists.txt:25]: Actually, the install directory is outside the source directory in the docker.

File: CMakeLists.txt
Comment (ddiakiteaneo) [reply CMakeLists.txt:1]: Step targets force CMake to build the dependency before continuing or performing a parallel build

File: ArmoniK.SDK.Client.Test/CMakeLists.txt (hunk 1 above)
Comment (ddiakiteaneo) [reply ArmoniK.SDK.Client.Test/CMakeLists.txt]: It's not needed but it simplifies the build when performing in two steps. I removed it

File: Utils.cmake [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    if(IS_ALPINE EQUAL -1)
+        message(STATUS "Not found Alpine distribution " ${IS_ALPINE})
+    else()
+        message(STATUS "Found Alpine distribution " ${IS_ALPINE})
+    endif()
+
+    if(CMAKE_BUILD_TYPE MATCHES DEBUG AND IS_ALPINE EQUAL -1)
+        target_link_options(${project_name_param} PRIVATE -fsanitize=undefined,address ${extra_param})
+    endif()
+endfunction()
```
Comment (dbrasseur-aneo) [inline Utils.cmake:29]:
```suggestion
endfunction()

```

**Review by dbrasseur-aneo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (ddiakiteaneo) [reply CMakeLists.txt:1]: That was a typo, fixed!

File: ArmoniK.SDK.DynamicWorker/CMakeLists.txt (hunk 6 above)
Comment (ddiakiteaneo) [reply ArmoniK.SDK.DynamicWorker/CMakeLists.txt:53]: Yes, there are totally equal!
The problem is when I use ${CMAKE_INSTALL_BINDIR} in install(FILES appsettings.json DESTINATION ${CMAKE_INSTALL_PREFIX}/bin) I will get an error if the bin directory is not yet created. The install file does not force the creation of the directory while the install target forces it.

**Review by dbrasseur-aneo — APPROVED** [review]

**Pull request merged**